### PR TITLE
propagate peer url to consensus if specified uri is does not match a …

### PR DESCRIPTION
**Pull request by generall**
…stored one + tests

test scenarios:

* create 3 nodes with replication factor 2 on 3 nodes
* update the collection
* kill random node
* Update the collection once again
* detect dead replicas
* delete dead peer from consensus
* Check functionality

----------

* create 3 nodes with replication factor 2 on 3 nodes
* update the collection
* kill random node
* Update the collection once again
* Start killed node without storage wipe
* wait until is syncs dead shards
* Check functionality

### CODE REVIEW

File: tests/consensus_tests/test_recover_dead_node.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    # 2 nodes majority should be enough for applying the status
+    create_collection(peer_api_uris[0], "test_collection2")
+
+    new_url = start_peer(peer_dirs[-1], f"peer_0_restarted.log", bootstrap_uri)
+
+    # Wait for cluster to recover dead peers
+    wait_for_all_replicas_active(peer_api_uris[0], "test_collection")
+
+    # Assert all records were changed
+    search_result = search(peer_api_uris[0], "Paris")
```
Comment (agourlay) [inline tests/consensus_tests/test_recover_dead_node.py:102]: What about additionally running the search query directly on the recovered node to control its state explicitly?

We could show that the replication is consistent regardless of the entry point.

```
search(peer_api_uris[0], "Paris").json["results"] == search(new_url, "Paris").json["results"]
```

File: tests/consensus_tests/test_remove_node.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                    "key": "city",
+                    "match": {"value": city}
+                }
+            ]
+        }
+    }
+    return requests.post(f"{peer_url}/collections/test_collection/points/search", json=q).json()["result"]
+
+
+def test_recover_dead_node(tmp_path: pathlib.Path):
```
Comment (agourlay) [inline tests/consensus_tests/test_remove_node.py:62]:
```suggestion
def test_remove_node(tmp_path: pathlib.Path):
```

File: tests/consensus_tests/test_remove_node.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    # Assert all records were changed
+    search_result = search(peer_api_uris[0], "Paris")
+    assert len(search_result) == 0
+
+    collection_cluster_info = get_collection_cluster_info(peer_api_uris[0], "test_collection")
+    dead_peer = [shard["peer_id"] for shard in collection_cluster_info["remote_shards"] if shard['state'] != 'Active']
+
+    assert len(dead_peer) > 1
+
+    # check new collection exists on a recovered node
```
Comment (agourlay) [inline tests/consensus_tests/test_remove_node.py:97]:
```suggestion
    # force remove dead peer
```

File: tests/consensus_tests/utils.py [hunk 4]
```diff
@@ -67,16 +67,19 @@ def get_qdrant_exec() -> str:
...
 
 # Starts a peer and returns its api_uri
-def start_peer(peer_dir: Path, log_file: str, bootstrap_uri: str) -> str:
-    p2p_port = get_port()
-    grpc_port = get_port()
-    http_port = get_port()
+def start_peer(peer_dir: Path, log_file: str, bootstrap_uri: str, port=None) -> str:
+    p2p_port = get_port() if port is None else port + 0
+    grpc_port = get_port() if port is None else port + 1
+    http_port = get_port() if port is None else port + 2
```
Comment (e-ivkov) [inline tests/consensus_tests/utils.py:73]: Cool! Didn't know about this `None` object

**Review by e-ivkov — APPROVED** [review]
LGTM

**Comment by agourlay** [comment]
There is an issue with the integration test after the `review fixes`

**Comment by generall** [comment]
> There is an issue with the integration test after the review fixes

great review ;)

**Pull request merged**